solutions/cses-1160.mdx

```diff
@@ -67,289 +35,114 @@ direction.
 using namespace std;
 
-using ll = long long;
-using ld = long double;
-using db = double;
-using str = string; // yay python!
-
-using pi = pair<int,int>;
-using pl = pair<ll,ll>;
-using pd = pair<db,db>;
-
-using vi = vector<int>;
-using vb = vector<bool>;
-using vl = vector<ll>;
-using vd = vector<db>;
-using vs = vector<str>;
-using vpi = vector<pi>;
-using vpl = vector<pl>;
-using vpd = vector<pd>;
-
-#define tcT template<class T
-#define tcTU tcT, class U
-// ^ lol this makes everything look weird but I'll try it
-tcT> using V = vector<T>;
-tcT, size_t SZ> using AR = array<T,SZ>;
-tcT> using PR = pair<T,T>;
-
-// pairs
-#define mp make_pair
-#define f first
-#define s second
-
-// vectors
-// oops size(x), rbegin(x), rend(x) need C++17
-#define sz(x) int((x).size())
-#define bg(x) begin(x)
-#define all(x) bg(x), end(x)
-#define rall(x) x.rbegin(), x.rend()
-#define sor(x) sort(all(x))
-#define rsz resize
-#define ins insert
-#define ft front()
-#define bk back()
-#define pb push_back
-#define eb emplace_back
-#define pf push_front
-
-#define lb lower_bound
-#define ub upper_bound
-tcT> int lwb(V<T>& a, const T& b) { return int(lb(all(a),b)-bg(a)); }
-
-// loops
-#define FOR(i,a,b) for (int i = (a); i < (b); ++i)
-#define F0R(i,a) FOR(i,0,a)
-#define ROF(i,a,b) for (int i = (b)-1; i >= (a); --i)
-#define R0F(i,a) ROF(i,0,a)
-#define trav(a,x) for (auto& a: x)
-
-const int MOD = 1e9+7; // 998244353;
-const int MX = 2e5+5;
-const ll INF = 1e18; // not too close to LLONG_MAX
-const ld PI = acos((ld)-1);
-const int dx[4] = {1,0,-1,0}, dy[4] = {0,1,0,-1}; // for every grid problem!!
-mt19937 rng((uint32_t)chrono::steady_clock::now().time_since_epoch().count());
-template<class T> using pqg = priority_queue<T,vector<T>,greater<T>>;
-
-// bitwise ops
-// also see https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html
-constexpr int pct(int x) { return __builtin_popcount(x); } // # of bits set
-constexpr int bits(int x) { // assert(x >= 0); // make C++11 compatible until USACO updates ...
-	return x == 0 ? 0 : 31-__builtin_clz(x); } // floor(log2(x))
-constexpr int p2(int x) { return 1<<x; }
-constexpr int msk2(int x) { return p2(x)-1; }
-
-ll cdiv(ll a, ll b) { return a/b+((a^b)>0&&a%b); } // divide a by b rounded up
-ll fdiv(ll a, ll b) { return a/b-((a^b)<0&&a%b); } // divide a by b rounded down
+const int MAXN = 2e5 + 1;
+const int LOGN = 20;
+
+int N, Q, cycle_len[MAXN];
+
+struct Node {
+	int par[LOGN], cycle_id = 0, dis = 0, vis = 0;
+	bool cycle = 0;
+} node[MAXN];
+
+pair<int, int> dfs(int u, int dfs_id) {
+	if (node[u].vis) {
+		if (node[u].vis == dfs_id) {  // cycle decomposition
+			int v = u, d = 0;
+			do {
+				node[v].cycle = 1;
+				node[v].cycle_id = u;
+				node[v].dis = d++;
+				v = node[v].par[0];
+			} while (v != u);
+			cycle_len[u] = d;
+			return {0, u};
+		} else {  // tree decomposition
+			if (node[u].cycle) {
+				return {node[u].cycle_id, 1};
+			} else {
+				return {node[u].cycle_id, node[u].dis + 1};
+			}
```

Comment:
```suggestion
			if (node[u].cycle) return {node[u].cycle_id, 1};
			else return {node[u].cycle_id, node[u].dis + 1};
```
#codecompress

Reply:
@SansPapyrus683

Reply:
back me up here

Reply:
aadit i'm going to fill your stomach with rocks